pootle/core/utils/version.py

```diff
@@ -0,0 +1,214 @@
+
+    return str(main + sub)
+
+
+def _is_development_candidate(version):
+    """Is this a pre-alpha release
+
+    >>> _is_development_candidate((2, 1, 0, 'alpha', 0))
+    True
+    >>> _is_development_candidate((2, 1, 0, 'beta', 1))
+    False
+    """
+    candidate_pos = _get_candidate_pos(version)
+    candidate = version[candidate_pos]
+    candidate_extra = version[candidate_pos+1]
+    return candidate == 'alpha' and candidate_extra == 0
+
+
+def _get_candidate_pos(version):
+    """Returns the position of the candidate marker.
+
+    >>> _get_candidate_pos((1, 2, 0, 'alpha', 0))
+    3
+    """
+    return [i for i, part in enumerate(version)
+            if part in CANDIDATE_MARKERS][0]
+
+
+def _get_candidate(version):
+    """Returns the candidate. One of alpha, beta, rc or final.
+
+    >>> _get_candidate((0, 1, 2, 'rc', 1))
+    'rc'
+    """
+    return version[_get_candidate_pos(version)]
+
+
+def _get_version_string(parts):
+    """Returns an X.Y.Z version from the list of version parts.
+
+    >>> _get_version_string((1, 1, 0))
+    '1.1.0'
+    >>> _get_version_string((1, 1, 0, 1))
+    '1.1.0.1'
+    """
+    return '.'.join(str(x) for x in parts)
+
+
+def get_main_version(version=None):
+    """Returns main version (X.Y[.Z]) from VERSION.
+
+    >>> get_main_version((1, 2, 3, 'alpha', 1))
+    '1.2.3'
+    """
+    version = get_complete_version(version)
+    candidate_pos = _get_candidate_pos(version)
+    return _get_version_string(version[:candidate_pos])
+
+
+def get_major_minor_version(version=None):
+    """Returns X.Y from VERSION.
+
+    >>> get_major_minor_version((1, 2, 3, 'final', 0))
+    '1.2'
+    """
+    version = get_complete_version(version)
+    return _get_version_string(version[:2])
+
+
+def get_complete_version(version=None):
+    """Returns a tuple of the Pootle version. Or the supplied ``version``
+
+    >>> get_complete_version()
+    (1, 2, 3, 'alpha', 0)
+    """
+    if version is None:
+        from pootle import VERSION as version
+
+    return version
+
+
+def get_docs_version(version=None, positions=2):
+    """Return the version used in documentation.
+
+    >>> get_docs_version((1, 2, 1, 'alpha', 0))
+    'dev'
+    >>> get_docs_version((1, 2, 1, 'rc', 2))
+    '1.2'
+    """
+    version = get_complete_version(version)
+    candidate = _get_candidate(version)
+    candidate_pos = _get_candidate_pos(version)
+    if positions > candidate_pos:
+        positions = candidate_pos
+    if _is_development_candidate(version):
+        return 'dev'
+    return _get_version_string(version[:positions])
+
+
+@lru_cache()
+def _shell_command(command):
+    """Return the first result of a shell ``command``"""
+    repo_dir = os.path.dirname(os.path.abspath(__file__))
+
+    command_subprocess = subprocess.Popen(
+        command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True,
+        cwd=repo_dir, universal_newlines=True
+    )
+
+    return command_subprocess.communicate()[0]
```

Comment:
Good spot.  First I thought it remote.  Then realised we'd rather not have code that can become a hole.  Slightly inelegant now but command specificed in the [] notation and lru_cache moved back to the calling functions, so we still can cache the results.